sqlanalytics/resource_query.go

```diff
@@ -0,0 +1,495 @@
+				}
+				if p.Enum.Multiple != nil {
+					p.Enum.Values = apv.Values
+				} else {
+					p.Enum.Value = apv.Values[0]
+				}
+			case *api.QueryParameterQuery:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.Query = &QueryParameterQuery{
+					QueryID:  apv.QueryID,
+					Multiple: newQueryParameterAllowMultiple(apv.Multi),
+				}
+				if p.Query.Multiple != nil {
+					p.Query.Values = apv.Values
+				} else {
+					p.Query.Value = apv.Values[0]
+				}
+			case *api.QueryParameterDate:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.Date = &QueryParameterDate{
+					Value: apv.Value,
+				}
+			case *api.QueryParameterDateTime:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.DateTime = &QueryParameterDateTime{
+					Value: apv.Value,
+				}
+			case *api.QueryParameterDateTimeSec:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.DateTimeSec = &QueryParameterDateTimeSec{
+					Value: apv.Value,
+				}
+			case *api.QueryParameterDateRange:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.DateRange = &QueryParameterDateRange{
+					Value: apv.Value,
+				}
+			case *api.QueryParameterDateTimeRange:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.DateTimeRange = &QueryParameterDateTimeRange{
+					Value: apv.Value,
+				}
+			case *api.QueryParameterDateTimeSecRange:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.DateTimeSecRange = &QueryParameterDateTimeSecRange{
+					Value: apv.Value,
+				}
+			default:
+				log.Fatalf("Don't know what to do for type: %#v", reflect.TypeOf(apv).String())
+			}
+
+			q.Parameter = append(q.Parameter, p)
+		}
+	}
+
+	// Transform to ResourceData.
+	if err := common.StructToData(*q, schema, data); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewQueryAPI ...
+func NewQueryAPI(ctx context.Context, m interface{}) QueryAPI {
+	return QueryAPI{m.(*common.DatabricksClient), ctx}
+}
+
+// QueryAPI ...
+type QueryAPI struct {
+	client  *common.DatabricksClient
+	context context.Context
+}
+
+func (a QueryAPI) buildPath(path ...string) string {
+	out := "/preview/sql/queries"
+	if len(path) == 1 {
+		out = out + "/" + strings.Join(path, "/")
+	}
+	return out
+}
+
+// Create ...
+func (a QueryAPI) Create(q *api.Query) (*api.Query, error) {
+	var qp api.Query
+	err := a.client.Post(a.context, a.buildPath(), q, &qp)
+	if err != nil {
+		return nil, err
+	}
+
+	// New queries are created with a table visualization by default.
+	// We don't manage that visualization here, so immediately remove it.
+	if len(qp.Visualizations) > 0 {
+		for _, rv := range qp.Visualizations {
+			var v api.Visualization
+			err = json.Unmarshal(rv, &v)
+			if err != nil {
+				return nil, err
+			}
+			// TODO
+			// err = a.DeleteVisualization(&v)
+			// if err != nil {
+			// 	return nil, err
+			// }
+		}
+		qp.Visualizations = []json.RawMessage{}
+	}
+
+	return &qp, err
+}
+
+// Read ...
+func (a QueryAPI) Read(q *api.Query) (*api.Query, error) {
```

Comment:
named return arguments would remove the need to cover error return statements here with tests ;)

Reply:
also.. it's very strange to Read with entire struct type to be passed. string ID must be enough to read a resource. (in terraform id is always a string)

Reply:
The sequence here was:
1. Need multiple IDs to read a visualization/widget.
2. Pass API object since it already captures both.
3. Use the same style everywhere and stick to it.

With the use of composite IDs in the latest commit, this is no longer needed, and read/update/delete take string parameters.